rfd/0007/readme.mdx

```diff
@@ -0,0 +1,157 @@
+title: Safe Database Schema Migration and Change Management
+state: discussion
+---
+
+## Abstract
+
+This RFD proposes a comprehensive approach to database schema migration and change management for Moose deployments
+that prioritizes safety, user control, and minimal operational overhead.
+The current solution lacks essential safeguards against unintended data loss and
+provides insufficient user control over schema evolution.
+
+The proposed solution implements a **Migration Plan Review and Approval** workflow where the system generates
+migration plans before deployment, allows user review and editing, and requires explicit approval before
+execution. This approach provides explicit user consent for destructive operations while maintaining
+flexibility for automated deployments in development environments.
+
+## Motivation / Problem Description
+
+The current schema migration system in Moose has several critical limitations that impact production deployments.
+
+### Current Process Issues
+
+In production mode, we load the infrastructure map stored in the docker image,
+update it according to the real database state,
+diff the changes, and apply them automatically.
+
+1. **Unintended Data Loss**: `DROP COLUMN` operations happen automatically without user awareness or approval.
+
+2. **No User Control**:
+   When users intend to rename a column,
+   the system interprets this as a `DROP COLUMN` followed by `ADD COLUMN`.
+   Data loss aside, the user has no way to specify how the schema is migrated.
+
+3. **State Inconsistency**:
+   The `integrate-changes` command modifies the infrastructure map,
+   but redeploying the same version means the stored infrastructure map in the image doesn't contain the new changes,
+   causing the system to drop tables as they're not in the desired state.
+
+4. **Async Cast Failures**:
+   `ALTER TABLE MODIFY COLUMN` operations don't change underlying data, and cast operations can fail asynchronously.
+
+### Goals
+
+- **Easy Migration**: Minimize work required for most common schema change scenarios
+- **User Control**: Provide full control and customization capabilities when needed  
+- **Safe Deployment**: Protect against unintended data loss during deployments and rollbacks
+- **Explicit Data Deletion**: Only drop data when the user explicitly requests it
+- **Rollback Safety**: Handle rollback scenarios without automatic data loss
+
+### Non-Goals
+
+- **Changing non-Moose managed database objects**: External tables, views, or procedures created outside of Moose
+
+## Solution: Migration Plan Review and Approval
+
+### Overview
+
+The plan is created and approved before deployment,
+and is not committed to the code base nor included in the container image.
+
+1. **Planning**: System generates migration plan before deployment
+2. **Review**:
+   - Runs image with `moose-cli plan` entry point, OR
+   - Send the target infrastructure map to running instance for plan generation
```

Comment:
How do we do this? Trying to understand if there's a way that can be fudged with by a bad actor along the way.